Allow optional block with Hash#delete

bmulvihill opened the pull request:
In Ruby we can pass an optional block to Hash#delete whose result can be returned if a key is not found.

This adds similar functionality to Crystal's Hash#delete.

Thanks!

Sija commented on src/hash.cr:
```diff
@@ -198,7 +198,23 @@ class Hash(K, V)
+    end
+  end
+
+  # Deletes the key-value pair and returns the value, else `nil`.
```
`else` -> `otherwise returns`

lbguilherme commented on spec/std/hash_spec.cr:
```diff
@@ -257,6 +257,15 @@ describe "Hash" do
+        a.delete(3) { found = false }
+        found.should be_false
+      end
+    end
```
You should also spec that the value of the block is returned. And that the method still finds a value if it exists.

lbguilherme commented:
Should `{5 => nil}.delete(5) { 3 }` return `3` or `nil`?

bmulvihill commented:
@lbguilherme It should return nil. I switched the implementation so it will only yield if the key is not found, so it now return nil, same as Ruby.

I also added 2 more specs as requested

Sija commented on spec/std/hash_spec.cr:
```diff
@@ -257,6 +257,25 @@ describe "Hash" do
+        a.delete(1).should eq(2)
+      end
+
+      it "executes block if key is not found" do
```
I'd test here case when found key has a `nil` value, like in `{ 3 => nil }.delete(3) { 7 }.should be_nil`

bmulvihill replied:
Done

asterite commented:
@bmulvihill Thank you!
